### ISSUE TITLE
rename adaptive_storage

### ISSUE THREAD

**Issue by HDembinski** [HDembinski/histogram#150]
adaptive_storage sounds too similar to storage_adaptor. The latter is difficult to rename, so I need to rename adaptive_storage. Options:
* dynamic_storage
* elastic_storage
* growing_storage

I slightly prefer elastic_storage.

**Comment by HDembinski** [comment]
@jpivarski @henryiii Any thoughts? Names are hard and input is appreciated!

**Comment by HDembinski** [comment]
adaptive_storage is the storage where the cell capacity grows as needed, so that it is able to storage larger and larger integers.

**Comment by jpivarski** [comment]
"Elastic" is hip (or was so recently; the name has died down a little). What your adaptive_storage does reminds me of "BigInt" (for computations) and "varint" (for serialization). It's not exactly either of these in that the whole array gets promoted (to maintain random access), but it's in the same spirit.

**Comment by HDembinski** [comment]
I like varint_storage, but it is not a perfect fit. The adaptive_storage does not only use integers, it is also switches to a double value when you scale it with a real number. That's why "growing_storage" is also not a perfect fit.

**Comment by henryiii** [comment]
I like `elastic_storage` best, I believe.

**Comment by henryiii** [comment]
Oh, wait, nevermind, I thought it was something else. I'll give input soon.

**Comment by jpivarski** [comment]
No, virtual_storage would be if it was lazily generated, which it's not. When I checked Wikipedia for "type erasure," it meant the reverse of "type inference," like `auto`. That's not what it is, either.

If it were me, I'd go with elastic. It's the best word proposed so far, in my opinion.

**Comment by henryiii** [comment]
Okay, I was thinking this was about axis, not storage. Here's my input:

* `elastic_storage`: This seems to be pretty good. As long as the reader doesn't miss the "storage" part, and understands this does not mean an axis will grow. `growing_storage` is similar, but has that problem even worse.
* `type_erasure_storage`: Wrong concept. This would mean I can mix types with something like `std::any`.
* `unlimited_storage`: This could work, but I assume you stop at 64 bits?
* `flexible_storage`: Might also work.
* `dynamic_storage`: Too similar to type_erasure, it sounds like you are able to put in different types and cast them.

**Comment by henryiii** [comment]
PS: Does each counter change as it hits its limit? I think `flexible_storage` or `flexible_numeric_storage` might be an option. Adding numeric makes it clear that this is just meant to change numeric types, but it probably too wordy.

**Comment by jpivarski** [comment]
Actually, I also like "flexible_storage". Maybe a little more than "elastic_storage" (because "elastic," as it was used in the past few years, meant starting up and shutting down servers to mean load demands). Smoothly transitioning from int8 through all integers to floating point is "flexible!"

**Comment by HDembinski** [comment]
Thank you guys, this is great input! So you guys prefer flexible_storage, which is a cool name. I like unlimited_storage. It is a powerful name for a powerful storage. The storage does not stop at 64 bits. If you are about to overflow 64 bits, the storage switches to a BigInt type, which is limited only by available memory. So, the name actually fits quite well. The marketing department of Boost.Histogram likes this choice.

**Comment by HDembinski** [comment]
And you are right, I did some more research yesterday, and type erasure does not quite fit.

**Comment by henryiii** [comment]
Given that it does switch to bigint, I like `unlimited_storage`. (I still like `flexible_storage` as well). `flexible_unlimited_numeric_storage` would be the most accurate description of what it does ;P.

**Comment by HDembinski** [comment]
True, but since you will not really interact with this storage a lot as a user, it is ok to have a less descriptive shorter name.

**Comment by HDembinski** [comment]
I also like flexible_storage, because the storage also switches to double if it is scaled by a real number.

**Comment by HDembinski** [comment]
I just think unlimited_storage sounds way cooler :)

**Comment by henryiii** [comment]
> That statement is backwards, IMO.

If you don't interact with something often, you need a longer, more descriptive name and you don't need to type as often so long is okay. `flexible_unlimited_storage` may not really be that bad...

**Comment by jpivarski** [comment]
My MINOS optimization has jumped to a new minimum twice: from adaptive_storage to flexible_storage, and now to unlimited_storage. Currently, I like unlimited_storage best. It's descriptive, doesn't conflict with a different concept, and explains why you ultimately want this.

**Comment by HDembinski** [comment]
@henryiii I see your point in general, but the concept of a storage is very well defined in the context of Boost.Histogram. And there are only two. There is no need to make the name more descriptive. And the users will not even touch the name. There is an alias, default_storage, which points to unlimited_storage. unlimited_storage is simply what you get if you do not explicitly want a different storage. Under these circumstances, I think the shorter name is admissible.
